frontend/src/hocs/withS4ACommunication/withS4ACommunication.tsx

```diff
@@ -0,0 +1,122 @@
+/**
+ * @license
+ * Copyright 2018-2020 Streamlit Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, { PureComponent, ComponentType, ReactNode } from "react"
+import hoistNonReactStatics from "hoist-non-react-statics"
+
+import {
+  IMenuItem,
+  IHostToGuestMessage,
+  IGuestToHostMessage,
+  VersionedMessage,
+} from "./types"
+
+interface State {
+  queryParams: string
+  items: IMenuItem[]
+}
+
+export interface S4ACommunicationHOC {
+  currentState: State
+  connect: () => void
+  sendMessage: (message: IGuestToHostMessage) => void
+}
+
+function withS4ACommunication(
+  WrappedComponent: ComponentType<any>
+): ComponentType<any> {
+  class ComponentWithS4ACommunication extends PureComponent<any, State> {
```

Comment:
Would it be easy/clearer to rewrite this as a function using React hooks? Up to you -- we're trying to avoid class components in new code, but perhaps this is clearer to understand?